routes/user.js

```diff
@@ -2,58 +2,72 @@
 
 const AppConstants = require("../app-constants");
 
-const crypto = require("crypto");
 const express = require("express");
 const router = express.Router();
 
 const EmailUtils = require("../email-utils");
 const Subscribers = require("../subscribers");
 
-// We send verification emails to addresses that want to subscribe.
-// These addresses are temporarily stored here, mapped to the unique
-// string token that is used for verification.
-const gUnverifiedEmails = new Map();
-
-router.post("/add", (req, res) => {
+router.post("/add", async (req, res) => {
   // TODO: use a hash of the email address instead of a random string.
-  const state = crypto.randomBytes(40).toString("hex");
+  const state = Subscribers.generateToken();
   req.session.state = state;
   const email = req.body.email;
   req.session.email = email;
-  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}`;
-
-  EmailUtils.sendEmail(email, "Firefox Breach Alert",
-    `Visit this link to subscribe: ${url}`)
-    .then(info => {
-      // TODO: set a timer to clear this after an arbitrary timeout period.
-      gUnverifiedEmails.set(state, email);
-      res.json({
-        email,
-        info: "sent verification link",
-        // Send the would-be link back to the client in dummy mode.
-        // eslint-disable-next-line no-process-env
-        link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
-      });
-    })
-    .catch(error => {
-      res.json({ email, info: "failed to send verification link", error });
-      throw error;
+  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}&email=${email}`;
```

Comment:
Sorry, i don't know the context of `state` or `email`, but do they need to be URI encoded?

Reply:
Good point thanks